File: MetroFuse-main/app/src/main/kotlin/com/metrolist/music/providers/IsrcResolver.kt
/**
 * Metrolist Project (C) 2026
 * Licensed under GPL-3.0 | See git history for contributors
 */

package com.metrolist.music.providers

import com.metrolist.music.apple.AppleMusicCanvasProvider
import com.metrolist.music.deezer.DeezerAudioProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap

/**
 * Resolves a trusted ISRC for a track from whichever source can supply one,
 * and (when a caller-supplied ISRC is already available) cross-checks it so
 * a bad tag can never silently poison an otherwise-correct match.
 *
 * Resolution order:
 *  1. Caller-supplied candidate ISRC (local tag / MediaStore / cached
 *     metadata / provider field) — normalized + validated shape only.
 *  2. Deezer catalog search by song+artist (cheap, no auth) — first result
 *     with a matching title/artist/duration wins.
 *  3. Apple Music catalog search by song+artist, using the same shared
 *     token as [AppleMusicCanvasProvider].
 *
 * Every lookup is cached in-memory keyed by (song, artist, durationSeconds)
 * so repeat plays of the same track never re-hit the network — this keeps
 * ISRC resolution off the hot path and safe to call from the UI-adjacent
 * playback pipeline.
 */
object IsrcResolver {

    private data class CacheKey(val song: String, val artist: String, val durationSeconds: Int?)

    // Positive + negative results both cached; NEGATIVE_RESULT means
    // "resolution was attempted and failed", distinct from "never attempted"
    // (absent key). ConcurrentHashMap disallows null values at the JVM
    // level (throws NPE from putVal), so a sentinel is used instead of null
    // even though the value type here is nullable.
    private const val NEGATIVE_RESULT = "\u0000NEGATIVE_RESULT\u0000"
    private val cache = ConcurrentHashMap<CacheKey, String>()

    /**
     * Returns a normalized, structurally valid ISRC, or null if none could
     * be resolved/validated. Never throws — network/parse failures degrade
     * to null so callers can fall through to lower-confidence matching.
     */
    suspend fun resolveAndValidate(
        candidateIsrc: String?,
        song: String,
        artist: String,
        durationSeconds: Int?,
    ): String? = withContext(Dispatchers.IO) {
        // 1. Caller-supplied candidate — normalize/validate only, no network.
        ProviderIsrc.normalize(candidateIsrc)?.let { return@withContext it }

        if (song.isBlank() || artist.isBlank()) return@withContext null

        val key = CacheKey(song.trim().lowercase(), artist.trim().lowercase(), durationSeconds)
        cache[key]?.let { return@withContext if (it == NEGATIVE_RESULT) null else it }

        val resolved = runCatching {
            coroutineScope {
                val deezerDeferred = async { resolveViaDeezer(song, artist, durationSeconds) }
                val appleDeferred = async { resolveViaApple(song, artist, durationSeconds) }
                deezerDeferred.await() ?: appleDeferred.await()
            }
        }.getOrNull()

        cache[key] = resolved ?: NEGATIVE_RESULT
        resolved
    }

    // Uses DeezerAudioProvider.findBestMatch, which runs the same
    // ISRC-first -> scored title/artist/album/duration search -> song.link
    // fallback chain used for actual playback resolution. This matters:
    // a naive "first search result that happens to carry an ISRC" can
    // easily grab a cover, remix, or same-titled track by a different
    // artist, which then silently poisons the Apple Music canvas match
    // (wrong ISRC -> wrong catalog track -> wrong canvas). Scoring against
    // title + artist + duration before accepting a candidate is what makes
    // this resolution path trustworthy enough to feed into ISRC-keyed
    // lookups elsewhere in the app.
    private suspend fun resolveViaDeezer(song: String, artist: String, durationSeconds: Int?): String? =
        runCatching {
            val query = DeezerAudioProvider.Query(
                mediaId = "",
                title = song,
                artists = listOf(artist),
                album = null,
                isrc = null,
                durationMs = durationSeconds?.toLong()?.times(1000L),
                resolverUrl = DeezerAudioProvider.DEFAULT_RESOLVER_URL,
                quality = com.metrolist.music.constants.DeezerAudioQuality.MP3_128,
                fastMode = false,
                proxyUrl = DeezerAudioProvider.DEFAULT_PROXY_URL,
            )
            DeezerAudioProvider.findBestMatch(query)
                ?.isrc
                ?.let { ProviderIsrc.normalize(it) }
        }.getOrNull()

    private suspend fun resolveViaApple(song: String, artist: String, durationSeconds: Int?): String? =
        runCatching {
            val token = AppleMusicCanvasProvider.borrowToken() ?: return@runCatching null
            AppleMusicCanvasProvider.searchIsrcOnly(song, artist, durationSeconds, token)
        }.getOrNull()

    /** Test/debug hook — clears the in-memory resolution cache. */
    fun clearCache() = cache.clear()
}

File: lib/services/providers/matching/isrc_resolver.dart
/// Resolves an ISRC for a track from metadata or a public catalog (Deezer),
/// tracking resolution provenance and enforcing timeouts.
library;

import 'dart:async';

import '../deezer/deezer_api.dart';
import '../song_query.dart';
import 'isrc.dart';
import 'track_candidate.dart';
import 'track_scorer.dart';

/// Provenance of an ISRC resolution.
enum IsrcProvenance {
  /// Provided directly by caller or track metadata (e.g. tags).
  provided,

  /// Resolved via public catalog (Deezer) search and verified by matching engine.
  catalogSearch,

  /// Retrieved from memory cache.
  cached,

  /// No ISRC could be established.
  none,
}

/// Detailed result of an ISRC resolution attempt.
class IsrcResolutionResult {
  final String? isrc;
  final IsrcProvenance provenance;
  final DateTime resolvedAt;
  final String? note;

  const IsrcResolutionResult({
    this.isrc,
    required this.provenance,
    required this.resolvedAt,
    this.note,
  });

  bool get isFound => isrc != null && isrc!.isNotEmpty;

  factory IsrcResolutionResult.none([String? note]) => IsrcResolutionResult(
        isrc: null,
        provenance: IsrcProvenance.none,
        resolvedAt: DateTime.now(),
        note: note,
      );

  factory IsrcResolutionResult.provided(String isrc) => IsrcResolutionResult(
        isrc: isrc,
        provenance: IsrcProvenance.provided,
        resolvedAt: DateTime.now(),
        note: 'Provided in source metadata',
      );

  factory IsrcResolutionResult.cached(String isrc) => IsrcResolutionResult(
        isrc: isrc,
        provenance: IsrcProvenance.cached,
        resolvedAt: DateTime.now(),
        note: 'Loaded from cache',
      );

  factory IsrcResolutionResult.catalog(String isrc) => IsrcResolutionResult(
        isrc: isrc,
        provenance: IsrcProvenance.catalogSearch,
        resolvedAt: DateTime.now(),
        note: 'Resolved via Deezer catalog search',
      );
}

class IsrcResolver {
  IsrcResolver({
    DeezerApi? deezerApi,
    Duration defaultTimeout = const Duration(seconds: 5),
  })  : _deezer = deezerApi ?? DeezerApi(),
        _timeout = defaultTimeout;

  final DeezerApi _deezer;
  final Duration _timeout;

  /// Cache of resolved ISRCs keyed by song/artist/duration. An empty string
  /// means "resolution attempted and failed" (distinct from absent).
  final Map<String, String> _cache = {};

  static const String _negative = '';

  /// Returns a normalized, structurally valid ISRC for the track, or null
  /// when none could be resolved. Never throws — failures degrade to null
  /// so callers fall through to lower-confidence matching.
  Future<String?> resolve({
    String? candidateIsrc,
    required String song,
    required String artist,
    int? durationMs,
  }) async {
    final result = await resolveWithProvenance(
      candidateIsrc: candidateIsrc,
      song: song,
      artist: artist,
      durationMs: durationMs,
    );
    return result.isrc;
  }

  /// Resolves an ISRC with provenance details.
  Future<IsrcResolutionResult> resolveWithProvenance({
    String? candidateIsrc,
    required String song,
    required String artist,
    int? durationMs,
  }) async {
    // 1. Caller-supplied ISRC — validate format first, no network needed
    final validated = normalizeIsrc(candidateIsrc);
    if (validated != null) {
      return IsrcResolutionResult.provided(validated);
    }

    if (song.trim().isEmpty || artist.trim().isEmpty) {
      return IsrcResolutionResult.none('Song title or artist is empty');
    }

    // 2. Cache lookup
    final key = '${song.trim().toLowerCase()}::${artist.trim().toLowerCase()}'
        '::$durationMs';
    final cached = _cache[key];
    if (cached != null) {
      if (cached.isEmpty) {
        return IsrcResolutionResult.none('Negative cache hit');
      }
      return IsrcResolutionResult.cached(cached);
    }

    // 3. Network lookup via Deezer catalog with timeout
    try {
      final resolved = await _resolveViaDeezer(song, artist, durationMs)
          .timeout(_timeout);
      _cache[key] = resolved ?? _negative;
      if (resolved != null) {
        return IsrcResolutionResult.catalog(resolved);
      }
      return IsrcResolutionResult.none('No matching catalog track with valid ISRC');
    } on TimeoutException {
      return IsrcResolutionResult.none('ISRC resolution timed out after ${_timeout.inSeconds}s');
    } catch (e) {
      return IsrcResolutionResult.none('ISRC resolution error: $e');
    }
  }

  /// Resolves ISRC for a [SongQuery]. If the query already has an ISRC,
  /// uses it directly without network call.
  Future<String?> resolveForQuery(SongQuery query) {
    return resolve(
      candidateIsrc: query.isrc,
      song: query.title,
      artist: query.artists.isNotEmpty ? query.artists.first : '',
      durationMs: query.durationMs,
    );
  }

  /// Searches Deezer's public catalog and returns the ISRC of the
  /// best-scoring track that actually matches [song]/[artist]/[duration].
  Future<String?> _resolveViaDeezer(
    String song,
    String artist,
    int? durationMs,
  ) async {
    try {
      final items = await _deezer.searchTracks('$artist - $song', limit: 12);
      if (items.isEmpty) return null;
      final candidates = <TrackCandidate>[];
      for (final item in items) {
        final title = item['title']?.toString();
        if (title == null || title.isEmpty) continue;
        candidates.add(TrackCandidate(
          trackId: item['id']?.toString() ?? '',
          title: title,
          artists: _artistNames(item),
          album: item['album'] is Map
              ? (item['album'] as Map)['title']?.toString()
              : null,
          isrc: normalizeIsrc(item['isrc']?.toString()),
          durationMs: item['duration'] is num
              ? ((item['duration'] as num) * 1000).round()
              : null,
        ));
      }
      final query = SongQuery(
        mediaId: '',
        title: song,
        artists: [artist],
        durationMs: durationMs,
      );
      final best = pickBestMatch(candidates, query);
      if (best == null) return null;
      return normalizeIsrc(best.isrc);
    } catch (_) {
      return null;
    }
  }

  List<String> _artistNames(Map<String, dynamic> item) {
    final names = <String>[];
    void addName(Object? name) {
      final value = name?.toString();
      if (value != null && value.isNotEmpty && !names.contains(value)) {
        names.add(value);
      }
    }

    final artist = item['artist'];
    if (artist is Map) addName(artist['name']);
    final contributors = item['contributors'];
    if (contributors is List) {
      for (final entry in contributors) {
        if (entry is Map) addName(entry['name']);
      }
    }
    return names;
  }
}
